### PR TITLE
provider/openstack: Refactored upsert load balancer to use lbaasv2

### PR DESCRIPTION
**Pull request by derekolk**

### CODE REVIEW

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/BlockingStatusChecker.groovy [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  final long pollInterval
+  final StatusChecker statusChecker
+
+  private BlockingStatusChecker(StatusChecker statusChecker, long timeout, long pollInterval) {
+    this.statusChecker = statusChecker
+    this.timeout = timeout
+    this.pollInterval = pollInterval
+  }
+
+  static BlockingStatusChecker from (StatusChecker s) {
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/BlockingStatusChecker.groovy]: Both `from` methods have an extra space before the `(`

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/BlockingStatusChecker.groovy [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import com.netflix.spinnaker.clouddriver.openstack.deploy.exception.OpenstackProviderException
+import groovy.transform.PackageScope
+
+import java.util.concurrent.TimeUnit
+
+/**
+ * This class abstracts the algorithm to continually poll until a status is obtained or a timeout occurs.  Openstack
+ * API requires that the load balancer be in an ACTIVE state for it to create associated relationships (i.e. listeners, pools,
+ * monitors).  Each modification will cause the load balancer to go into a PENDING state and back to ACTIVE once the change
+ * has been made.  Depending on your implementation, the timeout and polling intervals would need to be tweaked.
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/BlockingStatusChecker.groovy:28]: Are these default intervals intended to work across all of clouddriver?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/loadbalancer/UpsertOpenstackLoadBalancerAtomicOperation.groovy [hunk 3]
```diff
@@ -16,203 +16,390 @@
...
+        }
+      }
 
-  protected LbPool createLoadBalancer(String region, String subnetId, LoadBalancerPool newLoadBalancerPool, VirtualIP virtualIP, PoolHealthMonitor poolHealthMonitor) {
-    LbPool resultPool
-    OpenstackClientProvider openstackClientProvider = getClientProvider()
+      Set<String> listenersToDelete = Sets.difference(existingListenerMap.keySet(), Sets.union(listenersToAdd.keySet(), listenersToUpdate.keySet()))
 
-    newLoadBalancerPool.setCreatedTime(System.currentTimeMillis())
+      if (!listenersToDelete.isEmpty()) {
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/loadbalancer/UpsertOpenstackLoadBalancerAtomicOperation.groovy]: As far as I know, `isEmpty` is only really used when you have a reason to disambiguate between `null` and `[:]` or `[]`. I don't think this is the case here, so it's more idiomatic to rely on groovy-truthiness, e.g. `if (!listenersToDelete) {`

**Comment by lwander** [comment]
Looking good, nice work on making the upsert and delete operations so modular

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/BlockingStatusChecker.groovy (hunk 1 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/BlockingStatusChecker.groovy]: Will fix.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/BlockingStatusChecker.groovy (hunk 2 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/BlockingStatusChecker.groovy:28]: This is only used by load balancer now, but to your point, could be used in other areas.   I have decided to remove the from method with defaults and make it configuration based.  Therefore, the client using the checker will be responsible for providing timeout and interval values.  

Configuration will contain a reasonable default for stable installations,  Right now we are using devstack which is undersized and takes approximately 7 or 8 minutes to launch a load balancer.  These defaults could change as we get a stable liberty install.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/loadbalancer/UpsertOpenstackLoadBalancerAtomicOperation.groovy (hunk 3 above)
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/loadbalancer/UpsertOpenstackLoadBalancerAtomicOperation.groovy]: that is how I had it, changed it for an internal review ... Will revert back.

**Pull request closed without merging**